8320909: C2: Adapt IGVN's enqueuing logic to match idealization of AndNode with LShift operand

marc-chevalier opened the pull request:
The JBS issues has 3 reproducers. Two of them don't reproduce anymore. Let's enumerate:

- Test, TestSimple:
  Disappeared with:
  [JDK-8319372: C2 compilation fails with "Bad immediate dominator info"](https://bugs.openjdk.org/browse/JDK-8319372) in #16844
  which actually fixed it by removing the handling of the problematic pattern in `CastIINode::Value`.
  Reverting this fix makes the issue reappear.
- Reduced2: I fix here
- Test3, Reduced3:
  Disappeared with:
  [JDK-8347459: C2: missing transformation for chain of shifts/multiplications by constants](https://bugs.openjdk.org/browse/JDK-8347459) in #23728
  which just shadowed it. The bug is essentially the same as Reduced2 otherwise. I also fix these here (even with reverted JDK-8347459)

The issue comes from the fact that `And[IL]Node::Value` has a special handling when
an operand is a left-shift: in the expression
```
lhs & (X << s)
```
if `lhs` fits in less than `s` bits, the result is sure to be 0. This special handling
also tolerate a `ConvI2LNode` between the `AndLNode` and `LShiftINode`. In this case,
updating the Shift node during IGVN won't enqueue directly the And node, but only the
Conv node. If this conv node cannot be improved, the And node is not enqueued, and its
type is not as good as it could be.

Such a case is illustrated on the following figures from Reduced2. Node `239 Phi` is a phi with a
dead branch, so the node is about to be eleminated. On the second figure, we can see
`152 LShiftI` taking its place. The node `243 ConvI2L` is enqueued, but no change happens
during its idealization, so the node `244 AndL` is not enqueued, while it could receive an update.


<img src="https://github.com/user-attachments/assets/a68b27d0-b3cc-4850-b20e-0cff17047ed8" width="400">
<img src="https://github.com/user-attachments/assets/d98731e8-d85c-413d-a77e-059787b01884" width="400">

The fix is pretty direct: we recognize this pattern and we enqueue the And node during IGVN
to make sure it has a chance to be refined.

The case of Reduced3 is mostly the same, with a twist: the special handling of And nodes
also can see through casts. See the next figure with a `CastIINode` between the `AndINode` and
the `LShiftINode`. The fix has to take that into account.

<img src="https://github.com/user-attachments/assets/ae3d5a01-a0d6-4346-944c-4ba46af64ee2" width="400">


Overall, the situation can be of the form:
```
LShift -> Cast+ -> ConvI2L -> Cast+ -> And
```
This second case was shadowed by [JDK-8347459](https://bugs.openjdk.org/browse/JDK-8347459) because the structure
```
And(constant, Cast(LShift(LShift(X, 16), 16)))
```
that appears in the example is simplified into
```
And(constant, Cast(0))
```
which enqueue the Cast for IGVN, but in this form it can improve its type: it's surely 0,
so the And node is also processed to deduce the value 0 as well.

I tried and failed to create a reproducer of this case that isn't shadowed by [JDK-8347459](https://bugs.openjdk.org/browse/JDK-8347459); in particular changing the constant from 16 to 15, for instance, doesn't allow to simplify the double shift into 0, but doesn't reproduce the issue.

One could imagine an alternate solution that would not update the notification system
but rather the type system. On top of intervals, we could have some modulo information
and know that some value is 0 modulo 2^s, or alternatively, some bitwise information
that would know the lower `s` bits to be 0 (and the top bits to be unknown). This
would propagate through casts and convs and join through phis. This would allow the And
node to have best type earlier on. If the information is already there (like on the
first figure). Also, the And node would not need to look for deep structures, but only
compare the modulo/bitwise information of operands to know if the value can be refined,
which means the enqueuing code would not need to be changed, and more cases would
naturally work without special pattern-matching from And node. Nevertheless, this solution
feels way out-of-scope...

Thanks,
Marc

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320909](https://bugs.openjdk.org/browse/JDK-8320909): C2: Adapt IGVN's enqueuing logic to match idealization of AndNode with LShift operand (**Bug** - P4)


### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/24792/head:pull/24792` \
`$ git checkout pull/24792`

Update a local copy of the PR: \
`$ git checkout pull/24792` \
`$ git pull https://git.openjdk.org/jdk.git pull/24792/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24792`

View PR using the GUI difftool: \
`$ git pr show -t 24792`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/24792.diff">https://git.openjdk.org/jdk/pull/24792.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/24792#issuecomment-2820529176)
</details>

eme64 commented:
> One could imagine an alternate solution that would not update the notification system
but rather the type system. On top of intervals, we could have some modulo information
and know that some value is 0 modulo 2^s, or alternatively, some bitwise information
that would know the lower s bits to be 0 (and the top bits to be unknown).

@marc-chevalier FYI: https://github.com/openjdk/jdk/pull/17508 this could be an alternative solution, at least once we fully follow that road. It tracks the "known bits", which would give you the "modulo information".

eme64 commented on test/hotspot/jtreg/compiler/c2/gvn/MissingOptWithShiftConvAnd.java:
```diff
@@ -0,0 +1,65 @@
+ *          ConvI2L might not have an update when its input change. In this example, the
+ *          input was a Phi with a dead branch and becomes a LShiftI with the same type.
+ *
+ * @run main/othervm
+ *          -XX:CompileOnly=MissingOptWithShiftConvAnd::test
+ *          -XX:-TieredCompilation -Xbatch
+ *          -XX:+IgnoreUnrecognizedVMOptions -XX:VerifyIterativeGVN=10
+ *          MissingOptWithShiftConvAnd
```
It could be good to have a run without flags. Or at lest with fewer flags. Just so we can run it with other flag combinations. Imagine we run it with `-XX:VerifyIterativeGVN=11` from the outside, that would just get downgraded to `-XX:VerifyIterativeGVN=10` immediately here.

eme64 commented on test/hotspot/jtreg/compiler/c2/gvn/MissingOptWithShiftConvCastAnd.java:
```diff
@@ -0,0 +1,69 @@
+ * @summary Similar to MissingOptWithShiftConvAnd, but with Cast_II on the way.
+ * @library /test/lib
+ *
+ * @run main/othervm
+ *          -XX:CompileOnly=MissingOptWithShiftConvCastAnd::test
+ *          -XX:-TieredCompilation -Xbatch
+ *          -XX:+IgnoreUnrecognizedVMOptions -XX:VerifyIterativeGVN=10
+ *          MissingOptWithShiftConvCastAnd
```
Same here :)

eme64 reviewed (changes requested):
@marc-chevalier Thanks for looking into this! I think the solution looks good :)

Did you put in all 3 reproducers, even the ones that do not reproduce any more? It could be interesting if we ever backport, and maybe also more generally because it generates interesting patterns.

eme64 commented:
@marc-chevalier I also wonder if we could not have a more descriptive title?

BTW: **thank you** for the detailed PR description, very helpful :blush:

TobiHartmann commented:
> I just feared it would be confusing when browsing the history, to find a test actually fixed by another change than the one we will find in git blame.

I think that's fine - just make sure that you reference the right bug via the `@bug` tag in the test.

marc-chevalier commented:
I've added the other (not-anymore-)reproducer, tried to make clear where it comes from but where it was fixed. I've renamed the issue, added flag-free runs (because actually no flags are really needed for these). I think the PR is ready for more review.

eme64 commented on test/hotspot/jtreg/compiler/c2/gvn/MissedOptCastII.java:
```diff
@@ -0,0 +1,68 @@
+
+/*
+ * @test
+ * @bug 8319372
```
I would have just put both but ids here. Hope that is correct?
```suggestion
 * @bug 8319372 8320909
```

marc-chevalier replied:
Indeed, I found examples of that by grep. I had no idea that was possible! Done.

eme64 reviewed (approved):
@marc-chevalier thanks for the updates, looks good to me :)

marc-chevalier commented:
> @marc-chevalier FYI: https://github.com/openjdk/jdk/pull/17508 this could be an alternative solution, at least once we fully follow that road. It tracks the "known bits", which would give you the "modulo information".

Yes, exactly! A bitwise domain is exactly what we could use here. It'd give limited modulo/range information, but with bitwise operations, we care about bits. Godspeed to that! It's nice.

marc-chevalier commented:
/integrate

Thanks @eme64 and @TobiHartmann for reviews!

eme64 commented:
/sponsor
